README.md
[![Build Status](https://travis-ci.org/fivdi/mcp9808-temperature-sensor.svg?branch=master)](https://travis-ci.org/fivdi/mcp9808-temperature-sensor)
[![npm Version](http://img.shields.io/npm/v/mcp9808-temperature-sensor.svg)](https://www.npmjs.com/package/mcp9808-temperature-sensor)
[![Downloads Per Month](http://img.shields.io/npm/dm/mcp9808-temperature-sensor.svg)](https://www.npmjs.com/package/mcp9808-temperature-sensor)

# mcp9808-temperature-sensor

MCP9808 I2C temperature sensor module for Node.js on Linux boards like the
Raspberry Pi or BeagleBone.

Supports Node.js versions 10, 12, 14, 15 and 16.

## Contents

 * [Features](#features)
 * [Installation](#installation)
 * [Usage](#usage)
   * [Circuit](#circuit)
   * [Report the Temperature](#report-the-temperature)
   * [Alerts](#alerts)
 * [API](#api)
 * [Related Packages](#related-packages)

## Features

 * Simple temperature sensing
 * Alert detection
 * Real hardware interrupts used to detect alerts, no polling needed
 * Promise based asynchronous API

## Installation

```
npm install mcp9808-temperature-sensor
```

## Usage

#### Circuit

<img src="https://raw.githubusercontent.com/fivdi/mcp9808-temperature-sensor/master/examples/circuit.png">

Note that neither the 10kΩ pull-up resistor nor the wire connecting Alert on
the MCP9808 breakout board to GPIO27 on the Raspberry Pi are needed for the
"Report the Temperature" example. They are needed for the "Alerts" example.

#### Report the Temperature

Log the temperature in degrees Celsius to the console.

```js
const mcp9808 = require('mcp9808-temperature-sensor');

mcp9808.open().then(sensor =>
  sensor.temperature().
  then(temp => console.log(temp.celsius + '°C')).
  then(_ => sensor.close())
).catch(console.log);
```

#### Alerts

Fire alert events if the temperature drops below 25°C, goes above 35°C or
becomes critical at 45°C.

A hair dryer or blow dryer can be used to blow hot air over the MCP9808
temperature sensor and increase its temperature.

```js
const mcp9808 = require('mcp9808-temperature-sensor');

mcp9808.open({
  i2cBusNumber: 1, // optional, default 1
  i2cAddress: 0x18, // optional, default 0x18
  alertGpioNumber: 27,
  lowerAlertTemperature: 25,
  upperAlertTemperature: 35,
  criticalTemperature: 45
}).then(sensor => {
  setInterval(_ => {
    sensor.temperature().then(temp => console.log(temp.celsius + '°C'));
  }, 1000);

  sensor.on('alert', temp => {
    console.log('  alert ' + temp.celsius + '°C');
    if (temp.critical) {
      console.log('    critical');
    }
    if (temp.aboveUpperLimit) {
      console.log('    above upper limit');
    }
    if (temp.belowLowerLimit) {
      console.log('    below lower limit');
    }
  });

  return sensor.enableAlerts();
}).catch(console.log);
```

## API

### Functions

- [open([options])](#openoptions)

### Class Mcp9808

- [close()](#close)
- [temperature()](#temperature)
- [enableAlerts()](#enablealerts)

### Events

- [Event: 'alert'](#event-alert)
- [Event: 'error'](#event-error)

### Constants

- [RESOLUTION_1_2](#resolution_1_2)
- [RESOLUTION_1_4](#resolution_1_4)
- [RESOLUTION_1_8](#resolution_1_8)
- [RESOLUTION_1_16](#resolution_1_16)
- [HYSTERESIS_0](#hysteresis_0)
- [HYSTERESIS_1_5](#hysteresis_1_5)
- [HYSTERESIS_3](#hysteresis_3)
- [HYSTERESIS_6](#hysteresis_6)

#### open([options])
Returns a promise for an Mcp9808 object.

The following options are supported:
- i2cBusNumber - integer, I2C bus number, optional, default 1
- i2cAddress - integer, MCP9808 I2C address, optional, default 0x18
- alertGpioNumber - integer, GPIO number of the GPIO used for alert interrupt
detection, see [onoff](https://github.com/fivdi/onoff), optional, no default,
none used if none specified
- lowerAlertTemperature - number, alert lower temperature boundary, optional,
default 0°C
- upperAlertTemperature - number, alert upper temperature boundary, optional,
default 0°C
- criticalTemperature - number, alert critical temperature boundary, optional,
default 0°C
- resolution - integer, sensor resolution, see
[resolution constants](#resolution_1_2), optional, default 0.0625°C
- hysteresis - hysteresis for lower, upper and critical temperatures,
see [hysteresis constants](#hysteresis_0), optional, default 0°C

#### close()
Returns a promise which will resolve when all resources used by the Mcp9808
object have been freed.

#### temperature()
Returns a promise for an object containing
[temperature data](#temperature-data).

#### enableAlerts()
Returns a promise which will resolve when alerts have been enabled for the
Mcp9808.

### Event: 'alert'
- temperatureData - object containing [temperature data](#temperature-data)

Fired when an alert is detected.

### Event: 'error'
- err - Error object

Fired when an error occurs during alert detection.

### RESOLUTION_1_2
0.5°C resolution.

### RESOLUTION_1_4
0.25°C resolution.

### RESOLUTION_1_8
0.125°C resolution.

### RESOLUTION_1_16
0.0625°C resolution.

### HYSTERESIS_0
0°C hysteresis.

### HYSTERESIS_1_5
1.5°C hysteresis.

### HYSTERESIS_3
3°C hysteresis.

### HYSTERESIS_6
6°C hysteresis.

### Temperature Data
- celsius - number, temperature in degrees Celsius
- belowLowerLimit - boolean, true if temperature less than lower limit, else
false
- aboveUpperLimit - boolean, true if temperature greater than upper limit,
else false
- critical - boolean, true if temperature greater than or equal to critical
temperature, else false

## Related Packages

- [onoff](https://github.com/fivdi/onoff) - GPIO access and interrupt detection
- [i2c-bus](https://github.com/fivdi/i2c-bus) - I2C serial bus access



History.md
1.1.1 / Apr 29 2021
===================
  * drop support for node.js 8 and 13, add support for node.js 15 & 16
  * update dependencies

1.1.0 / Apr 25 2020
===================
  * update dependencies (i2c-bus v5.2.0, onoff v6.0.0, jshint v2.11.0)
  * drop support for node.js 6, add support for node.js 14

1.0.13 / Sep 23 2019
====================
  * update dependencies (onoff v5.0.0)

1.0.12 / Sep 21 2019
====================
  * document error event

1.0.11 / Sep 21 2019
====================
  * drop support for node 4
  * update dependencies (i2c-bus v5.0.0)
  * improve code style
  * refactor open function

1.0.10 / Sep 07 2019
====================
  * update dependencies (i2c-bus v4.0.11, onoff v4.1.4)

1.0.9 / Jun 16 2019
===================
  * update dependencies
  * update npm keywords
  * remove node 11 from build

1.0.8 / Mar 17 2019
===================
  * lint with jshint
  * add travis build
  * add badges

1.0.7 / Mar 14 2019
===================

  * update dependencies (i2c-bus v4.0.9, onoff v4.1.1)

1.0.6 / Sep 30 2018
===================

  * update dependencies (i2c-bus v4.0.2, onoff v3.2.2)

1.0.5 / Jul 25 2018
===================

  * update dependencies (i2c-bus v4.0.0, onoff v3.2.0)

1.0.4 / Apr 07 2018
===================

  * update dependencies (i2c-bus v3.1.0, onoff v3.0.2)

1.0.3 / Feb 26 2018
===================

  * update dependencies (onoff v2.0.0)

1.0.2 / Feb 25 2018
===================

  * update dependencies (i2c-bus v3.0.0)

1.0.1 / Feb 20 2018
===================

  * add alert detection tests

1.0.0 / Feb 18 2018
===================

  * initial release



test/open-errors.js
'use strict';

const assert = require('assert');
const mcp9808 = require('../');

// options
mcp9808.open('not an options object').
then(sensor => assert(false, 'Expected open to reject invalid options object.')).
catch(err => {
  if (err.message !== 'Expected options to be of type object. Got type string.') {
    console.log(err.stack);
  }
});

// i2cBusNumber
mcp9808.open({i2cBusNumber: 'a string'}).
then(sensor => assert(false, 'Expected open to reject invalid i2cBusNumber.')).
catch(err => {
  if (err.message !== 'Expected i2cBusNumber to be a non-negative integer. Got "a string".') {
    console.log(err.stack);
  }
});

mcp9808.open({i2cBusNumber: -1}).
then(sensor => assert(false, 'Expected open to reject invalid i2cBusNumber.')).
catch(err => {
  if (err.message !== 'Expected i2cBusNumber to be a non-negative integer. Got "-1".') {
    console.log(err.stack);
  }
});

mcp9808.open({i2cBusNumber: 1.5}).
then(sensor => assert(false, 'Expected open to reject invalid i2cBusNumber.')).
catch(err => {
  if (err.message !== 'Expected i2cBusNumber to be a non-negative integer. Got "1.5".') {
    console.log(err.stack);
  }
});

// i2cAddress
mcp9808.open({i2cAddress: 'a string'}).
then(sensor => assert(false, 'Expected open to reject invalid i2cAddress.')).
catch(err => {
  if (err.message !== 'Expected i2cAddress to be an integer >= 0 and <= 0x7f. Got "a string".') {
    console.log(err.stack);
  }
});

mcp9808.open({i2cAddress: -1}).
then(sensor => assert(false, 'Expected open to reject invalid i2cAddress.')).
catch(err => {
  if (err.message !== 'Expected i2cAddress to be an integer >= 0 and <= 0x7f. Got "-1".') {
    console.log(err.stack);
  }
});

mcp9808.open({i2cAddress: 128}).then(sensor =>
  assert(false, 'Expected open to reject invalid i2cAddress.')
).catch(err => {
  if (err.message !== 'Expected i2cAddress to be an integer >= 0 and <= 0x7f. Got "128".') {
    console.log(err.stack);
  }
});

mcp9808.open({i2cAddress: 1.5}).
then(sensor => assert(false, 'Expected open to reject invalid i2cAddress.')).
catch(err => {
  if (err.message !== 'Expected i2cAddress to be an integer >= 0 and <= 0x7f. Got "1.5".') {
    console.log(err.stack);
  }
});

// alertGpioNumber
mcp9808.open({alertGpioNumber: 'a string'}).
then(sensor => assert(false, 'Expected open to reject invalid alertGpioNumber.')).
catch(err => {
  if (err.message !== 'Expected alertGpioNumber to be a non-negative integer. Got "a string".') {
    console.log(err.stack);
  }
});

mcp9808.open({alertGpioNumber: -1}).
then(sensor => assert(false, 'Expected open to reject invalid alertGpioNumber.')).
catch(err => {
  if (err.message !== 'Expected alertGpioNumber to be a non-negative integer. Got "-1".') {
    console.log(err.stack);
  }
});

mcp9808.open({alertGpioNumber: 1.5}).
then(sensor => assert(false, 'Expected open to reject invalid alertGpioNumber.')).
catch(err => {
  if (err.message !== 'Expected alertGpioNumber to be a non-negative integer. Got "1.5".') {
    console.log(err.stack);
  }
});

// lowerAlertTemperature
mcp9808.open({lowerAlertTemperature: 'a string'}).
then(sensor => assert(false, 'Expected open to reject invalid lowerAlertTemperature.')).
catch(err => {
  if (err.message !== 'Expected lowerAlertTemperature to be a number >= -256 and <= 255.75. Got "a string".') {
    console.log(err.stack);
  }
});

mcp9808.open({lowerAlertTemperature: -300}).
then(sensor => assert(false, 'Expected open to reject invalid lowerAlertTemperature.')).
catch(err => {
  if (err.message !== 'Expected lowerAlertTemperature to be a number >= -256 and <= 255.75. Got "-300".') {
    console.log(err.stack);
  }
});

mcp9808.open({lowerAlertTemperature: 300}).
then(sensor => assert(false, 'Expected open to reject invalid lowerAlertTemperature.')).
catch(err => {
  if (err.message !== 'Expected lowerAlertTemperature to be a number >= -256 and <= 255.75. Got "300".') {
    console.log(err.stack);
  }
});

// upperAlertTemperature
mcp9808.open({upperAlertTemperature: 'a string'}).
then(sensor => assert(false, 'Expected open to reject invalid upperAlertTemperature.')).
catch(err => {
  if (err.message !== 'Expected upperAlertTemperature to be a number >= -256 and <= 255.75. Got "a string".') {
    console.log(err.stack);
  }
});

mcp9808.open({upperAlertTemperature: -400}).
then(sensor => assert(false, 'Expected open to reject invalid upperAlertTemperature.')).
catch(err => {
  if (err.message !== 'Expected upperAlertTemperature to be a number >= -256 and <= 255.75. Got "-400".') {
    console.log(err.stack);
  }
});

mcp9808.open({upperAlertTemperature: 400}).
then(sensor => assert(false, 'Expected open to reject invalid upperAlertTemperature.')).
catch(err => {
  if (err.message !== 'Expected upperAlertTemperature to be a number >= -256 and <= 255.75. Got "400".') {
    console.log(err.stack);
  }
});

// criticalTemperature
mcp9808.open({criticalTemperature: 'a string'}).
then(sensor => assert(false, 'Expected open to reject invalid criticalTemperature.')).
catch(err => {
  if (err.message !== 'Expected criticalTemperature to be a number >= -256 and <= 255.75. Got "a string".') {
    console.log(err.stack);
  }
});

mcp9808.open({criticalTemperature: -500}).
then(sensor => assert(false, 'Expected open to reject invalid criticalTemperature.')).
catch(err => {
  if (err.message !== 'Expected criticalTemperature to be a number >= -256 and <= 255.75. Got "-500".') {
    console.log(err.stack);
  }
});

mcp9808.open({criticalTemperature: 500}).
then(sensor => assert(false, 'Expected open to reject invalid criticalTemperature.')).
catch(err => {
  if (err.message !== 'Expected criticalTemperature to be a number >= -256 and <= 255.75. Got "500".') {
    console.log(err.stack);
  }
});

// resolution
mcp9808.open({resolution: 'a string'}).
then(sensor => assert(false, 'Expected open to reject invalid resolution.')).
catch(err => {
  if (err.message !== 'Expected resolution to be an integer >= 0 and <= 3. Got "a string".') {
    console.log(err.stack);
  }
});

mcp9808.open({resolution: -1}).
then(sensor => assert(false, 'Expected open to reject invalid resolution.')).
catch(err => {
  if (err.message !== 'Expected resolution to be an integer >= 0 and <= 3. Got "-1".') {
    console.log(err.stack);
  }
});

mcp9808.open({resolution: 4}).
then(sensor => assert(false, 'Expected open to reject invalid resolution.')).
catch(err => {
  if (err.message !== 'Expected resolution to be an integer >= 0 and <= 3. Got "4".') {
    console.log(err.stack);
  }
});

// hysteresis
mcp9808.open({hysteresis: 'a string'}).
then(sensor => assert(false, 'Expected open to reject invalid hysteresis.')).
catch(err => {
  if (err.message !== 'Expected hysteresis to be an integer >= 0 and <= 3. Got "a string".') {
    console.log(err.stack);
  }
});

mcp9808.open({hysteresis: -1}).
then(sensor => assert(false, 'Expected open to reject invalid hysteresis.')).
catch(err => {
  if (err.message !== 'Expected hysteresis to be an integer >= 0 and <= 3. Got "-1".') {
    console.log(err.stack);
  }
});

mcp9808.open({hysteresis: 4}).
then(sensor => assert(false, 'Expected open to reject invalid hysteresis.')).
catch(err => {
  if (err.message !== 'Expected hysteresis to be an integer >= 0 and <= 3. Got "4".') {
    console.log(err.stack);
  }
});

// alert temperatures
mcp9808.open({lowerAlertTemperature: 20}).
then(sensor => assert(false, 'Expected open to reject if correct number of alert temperatures not specified.')).
catch(err => {
  if (err.message !== 'Expected all alert temperatures or no alert temperatures to be specified.') {
    console.log(err.stack);
  }
});

mcp9808.open({lowerAlertTemperature: 20, upperAlertTemperature: 30}).
then(sensor => assert(false, 'Expected open to reject if correct number of alert temperatures not specified.')).
catch(err => {
  if (err.message !== 'Expected all alert temperatures or no alert temperatures to be specified.') {
    console.log(err.stack);
  }
});

mcp9808.open({upperAlertTemperature: 30, criticalTemperature: 40}).
then(sensor => assert(false, 'Expected open to reject if correct number of alert temperatures not specified.')).
catch(err => {
  if (err.message !== 'Expected all alert temperatures or no alert temperatures to be specified.') {
    console.log(err.stack);
  }
});

mcp9808.open({lowerAlertTemperature: 30, upperAlertTemperature: 30, criticalTemperature: 40}).
then(sensor => assert(false, 'Expected open to reject bad temperatures.')).
catch(err => {
  if (err.message !== 'Expected lowerAlertTemperature to be < upperAlertTemperature.') {
    console.log(err.stack);
  }
});

mcp9808.open({lowerAlertTemperature: 20, upperAlertTemperature: 40, criticalTemperature: 40}).
then(sensor => assert(false, 'Expected open to reject bad temperatures.')).
catch(err => {
  if (err.message !== 'Expected upperAlertTemperature to be < criticalTemperature.') {
    console.log(err.stack);
  }
});




test/open-read-close.js
'use strict';

const assert = require('assert');
const mcp9808 = require('../');

mcp9808.open().then(sensor =>
  sensor.temperature().
  then(temperature => console.log('  temperature = ' + temperature.celsius + '°C')).
  then(_ => sensor.close())
).catch(console.log);



test/detect-no-alerts.js
'use strict';

const mcp9808 = require('../');
const currentTemperature = require('./current-temperature');

currentTemperature().then(currentTemp =>
  mcp9808.open({
    alertGpioNumber: 27,
    lowerAlertTemperature: currentTemp - 10,
    upperAlertTemperature: currentTemp + 10,
    criticalTemperature: currentTemp + 20
  })
).then(sensor => {
  let watchdog = setTimeout(_ => {
    throw new Error('Expected an alert event with no alerts.');
  }, 1000);

  sensor.on('alert', temperature => {
    if (!temperature.belowLowerLimit &&
        !temperature.aboveUpperLimit &&
        !temperature.critical) {
      clearTimeout(watchdog);
      setTimeout(_ => sensor.close().catch(console.log), 100);
    }    
  });

  return sensor.enableAlerts();
}).catch(console.log);



mcp9808-temperature-sensor.js
'use strict';

const EventEmitter = require('events');
const Gpio = require('onoff').Gpio;
const i2c = require('i2c-bus');
const mutexify = require('mutexify');

const DEFAULT_I2C_BUS = 1;
const DEFAULT_I2C_ADDRESS = 0x18;

const MANUFACTURER_ID = 0x54;
const DEVICE_ID = 0x04;

const CONFIGURATION_REG = 0x01;
const UPPER_ALERT_TEMP_REG = 0x02;
const LOWER_ALERT_TEMP_REG = 0x03;
const CRITICAL_TEMP_REG = 0x04;
const TEMP_REG = 0x05;
const MANUFACTURER_ID_REG = 0x06;
const DEVICE_ID_REVISION_REG = 0x07;
const RESOLUTION_REG = 0x08;

const MIN_TEMPERATURE = -256;
const MAX_TEMPERATURE = 0xffc / 16; // 255.75

const MIN_RESOLUTION = 0;
const MAX_RESOLUTION = 3;

const MIN_HYSTERESIS = 0;
const MAX_HYSTERESIS = 3;

const CONFIGURATION_BITS = 0x7ff;

const HYSTERESIS_BITS = 0x0600;
const SHUTDOWN_BIT = 0x0100;
const INTERRUPT_CLEAR_BIT = 0x0020;
const ALERT_ENABLED_BIT = 0x0008;
const ALERT_MODE_BIT = 0x0001;

const configRegLock = mutexify();

const validateOpenOptions = options => {
  if (typeof options !== 'object') {
    return 'Expected options to be of type object.' +
      ' Got type ' + typeof options + '.';
  }

  if (options.i2cBusNumber !== undefined &&
      (!Number.isSafeInteger(options.i2cBusNumber) ||
       options.i2cBusNumber < 0
      )
     ) {
    return 'Expected i2cBusNumber to be a non-negative integer.' +
      ' Got "' + options.i2cBusNumber + '".';
  }

  if (options.i2cAddress !== undefined &&
      (!Number.isSafeInteger(options.i2cAddress) ||
       options.i2cAddress < 0 ||
       options.i2cAddress > 0x7f
      )
     ) {
    return 'Expected i2cAddress to be an integer' +
      ' >= 0 and <= 0x7f. Got "' + options.i2cAddress + '".';
  }

  if (options.alertGpioNumber !== undefined &&
      (!Number.isSafeInteger(options.alertGpioNumber) ||
       options.alertGpioNumber < 0
      )
     ) {
    return 'Expected alertGpioNumber to be a non-negative integer. ' +
      'Got "' + options.alertGpioNumber + '".';
  }

  if (options.lowerAlertTemperature !== undefined &&
      (typeof options.lowerAlertTemperature !== 'number' ||
       options.lowerAlertTemperature < MIN_TEMPERATURE ||
       options.lowerAlertTemperature > MAX_TEMPERATURE
      )
     ) {
    return 'Expected lowerAlertTemperature to be a number >= ' +
      MIN_TEMPERATURE + ' and <= ' + MAX_TEMPERATURE +
      '. Got "' + options.lowerAlertTemperature + '".';
  }

  if (options.upperAlertTemperature !== undefined &&
      (typeof options.upperAlertTemperature !== 'number' ||
       options.upperAlertTemperature < MIN_TEMPERATURE ||
       options.upperAlertTemperature > MAX_TEMPERATURE
      )
     ) {
    return 'Expected upperAlertTemperature to be a number >= ' +
      MIN_TEMPERATURE + ' and <= ' + MAX_TEMPERATURE +
      '. Got "' + options.upperAlertTemperature + '".';
  }

  if (options.criticalTemperature !== undefined &&
      (typeof options.criticalTemperature !== 'number' ||
       options.criticalTemperature < MIN_TEMPERATURE ||
       options.criticalTemperature > MAX_TEMPERATURE
      )
     ) {
    return 'Expected criticalTemperature to be a number >= ' +
      MIN_TEMPERATURE + ' and <= ' + MAX_TEMPERATURE +
      '. Got "' + options.criticalTemperature + '".';
  }

  if (options.criticalTemperature !== undefined &&
      (typeof options.criticalTemperature !== 'number' ||
       options.criticalTemperature < MIN_TEMPERATURE ||
       options.criticalTemperature > MAX_TEMPERATURE
      )
     ) {
    return 'Expected criticalTemperature to be a number >= ' +
      MIN_TEMPERATURE + ' and <= ' + MAX_TEMPERATURE +
      '. Got "' + options.criticalTemperature + '".';
  }

  if (options.resolution !== undefined &&
      (!Number.isSafeInteger(options.resolution) ||
       options.resolution < MIN_RESOLUTION ||
       options.resolution > MAX_RESOLUTION
      )
     ) {
    return 'Expected resolution to be an integer >= ' +
      MIN_RESOLUTION + ' and <= ' + MAX_RESOLUTION +
      '. Got "' + options.resolution + '".';
  }

  if (options.hysteresis !== undefined &&
      (!Number.isSafeInteger(options.hysteresis) ||
       options.hysteresis < MIN_HYSTERESIS ||
       options.hysteresis > MAX_HYSTERESIS
      )
     ) {
    return 'Expected hysteresis to be an integer >= ' +
      MIN_HYSTERESIS + ' and <= ' + MAX_HYSTERESIS +
      '. Got "' + options.hysteresis + '".';
  }

  if (options.lowerAlertTemperature !== undefined ||
      options.upperAlertTemperature !== undefined ||
      options.criticalTemperature !== undefined) {

    if (options.lowerAlertTemperature === undefined ||
        options.upperAlertTemperature === undefined ||
        options.criticalTemperature === undefined) {
      return 'Expected all alert temperatures' +
        ' or no alert temperatures to be specified.';
    }

    let lowerAlertTemperature =
      Math.round(options.lowerAlertTemperature * 4) / 4;
    let upperAlertTemperature =
      Math.round(options.upperAlertTemperature * 4) / 4;
    let criticalTemperature =
      Math.round(options.criticalTemperature * 4) / 4;

    if (lowerAlertTemperature >= upperAlertTemperature) {
      return 'Expected lowerAlertTemperature to be < upperAlertTemperature.';
    }

    if (upperAlertTemperature >= criticalTemperature) {
      return 'Expected upperAlertTemperature to be < criticalTemperature.';
    }
  }

  return null;
};

const watchAlertGpio = (alertGpio, tempSensor, i2cMcp9808) => {
  let lastAlertWasCritical = false;

  alertGpio.watch((err, value) => {
    if (err) {
      tempSensor.emit('error', err);
      return;
    }

    let fallingEdge = (value === 0);

    tempSensor.temperature().then(temperature => {
      if (fallingEdge || lastAlertWasCritical) {
        tempSensor.emit('alert', temperature);
      }

      lastAlertWasCritical = temperature.critical;

      if (fallingEdge && !temperature.critical) {
        return i2cMcp9808.configuration(INTERRUPT_CLEAR_BIT, 0);
      }
    }).catch(err => tempSensor.emit('error', err));
  });
};

class I2cMcp9808 {
  constructor(i2cBus, i2cAddress) {
    this._i2cBus = i2cBus;
    this._i2cAddress = i2cAddress;
  }

  close() {
    return this._i2cBus.close();
  }

  writeByte(register, byte) {
    return this._i2cBus.writeByte(this._i2cAddress, register, byte);
  }

  readWord(register) {
    return this._i2cBus.readWord(this._i2cAddress, register).
      then(word => (word >> 8) + ((word & 0xff) << 8));
  }

  writeWord(register, word) {
    return this._i2cBus.writeWord(
      this._i2cAddress, register, (word >> 8) + ((word & 0xff) << 8)
    );
  }

  softReset() {
    return this.configuration(0, SHUTDOWN_BIT).
      then(_ => this.configuration(0, ALERT_ENABLED_BIT)).
      then(_ => this.configuration(0, ALERT_MODE_BIT)). // TODO - can this be removed?
      then(_ => this.lowerAlertTemperature(0)).
      then(_ => this.upperAlertTemperature(0)).
      then(_ => this.criticalTemperature(0)).
      then(_ => this.resolution(Mcp9808.RESOLUTION_1_16)).
      then(_ =>
        this.configuration(
          INTERRUPT_CLEAR_BIT, CONFIGURATION_BITS ^ INTERRUPT_CLEAR_BIT
        )
      );
  }

  configuration(bitsToSet, bitsToReset) {
    let releaseConfigRegLock = null;

    // To modify bits in the configuration register it's necessary to read
    // the register, modify the required bits and write back to the
    // register. In order to prevent parallel asynchronous operations that
    // are modifying the configuration register from stepping on each other
    // here a mutex is needed.
    return new Promise((resolve, reject) => {
      configRegLock(release => {
        releaseConfigRegLock = release;
        resolve();
      });
    }).
    then(_ => this.readWord(CONFIGURATION_REG)).
    then(config => {
      config |= bitsToSet;
      config &= (~bitsToReset & 0xffff);

      return this.writeWord(CONFIGURATION_REG, config);
    }).
    then(_ => releaseConfigRegLock()).
    catch(err => {
      if (releaseConfigRegLock !== null) {
        releaseConfigRegLock();
      }
      return Promise.reject(err);
    });
  }

  writeTemperature(tempRegister, temp) {
    return Promise.resolve().then(_ => {
      let rawTemp = temp < 0 ? temp + 256 : temp;
      rawTemp = (Math.round(rawTemp * 4) << 2);

      if (temp < 0) {
        rawTemp |= 0x1000;
      }

      return this.writeWord(tempRegister, rawTemp);
    });
  }

  temperature() {
    return this.readWord(TEMP_REG).then(rawTemp => {
      let temp = (rawTemp & 0x0fff) / 16;

      if (rawTemp & 0x1000) {
        temp -= 256;
      }

      return {
        celsius: temp,
        belowLowerLimit: !!(rawTemp & 0x2000),
        aboveUpperLimit: !!(rawTemp & 0x4000),
        critical: !!(rawTemp & 0x8000),
        rawTemp: rawTemp
      };
    });
  }

  lowerAlertTemperature(temp) {
    return this.writeTemperature(LOWER_ALERT_TEMP_REG, temp);
  }

  upperAlertTemperature(temp) {
    return this.writeTemperature(UPPER_ALERT_TEMP_REG, temp);
  }

  criticalTemperature(temp) {
    return this.writeTemperature(CRITICAL_TEMP_REG, temp);
  }

  manufacturerId() {
    return this.readWord(MANUFACTURER_ID_REG);
  }

  deviceId() {
    return this.readWord(DEVICE_ID_REVISION_REG).then(word => word >> 8);
  }

  resolution(val) {
    return this.writeByte(RESOLUTION_REG, val);
  }

  hysteresis(val) {
    return this.configuration(
      (val << 9) & HYSTERESIS_BITS, (~val << 9) & HYSTERESIS_BITS
    );
  }

  enableAlerts() {
    // Don't set alertEnabled and alertMode at the same time. If done, there
    // will be no interrupt if the application is started and the current
    // temperature is > upperAlertTemperature and < criticalTemperature.
    return this.configuration(ALERT_ENABLED_BIT, 0).
      then(_ => this.configuration(ALERT_MODE_BIT, 0));
  }
}

class Mcp9808 extends EventEmitter {
  static get RESOLUTION_1_2() {return 0;}
  static get RESOLUTION_1_4() {return 1;}
  static get RESOLUTION_1_8() {return 2;}
  static get RESOLUTION_1_16() {return 3;}

  static get HYSTERESIS_0() {return 0;}
  static get HYSTERESIS_1_5() {return 1;}
  static get HYSTERESIS_3() {return 2;}
  static get HYSTERESIS_6() {return 3;}

  constructor(i2cMcp9808, alertGpio) {
    super();

    this._i2cMcp9808 = i2cMcp9808;
    this._alertGpio = alertGpio;
  }

  static open(options) {
    let i2cMcp9808;
    let tempSensor;

    return Promise.resolve().then(_ => {
      options = options || {};

      let errMsg = validateOpenOptions(options);
      if (errMsg) {
        return Promise.reject(new Error(errMsg));
      }

      return i2c.openPromisified(options.i2cBusNumber === undefined ?
        DEFAULT_I2C_BUS : options.i2cBusNumber);
    }).then(i2cBus => {
      const i2cAddress = options.i2cAddress === undefined ?
        DEFAULT_I2C_ADDRESS : options.i2cAddress;
      i2cMcp9808 = new I2cMcp9808(i2cBus, i2cAddress);

      let alertGpio = null;
      if (options.alertGpioNumber !== undefined) {
        alertGpio = new Gpio(options.alertGpioNumber, 'in', 'both');
      }

      tempSensor = new Mcp9808(i2cMcp9808, alertGpio);

      if (alertGpio !== null) {
        watchAlertGpio(alertGpio, tempSensor, i2cMcp9808);
      }
    }).then(_ =>
      i2cMcp9808.manufacturerId()
    ).then(manufacturerId => {
      if (manufacturerId !== MANUFACTURER_ID) {
        return Promise.reject(new Error(
          'Expected manufacturer ID to be 0x' + MANUFACTURER_ID.toString(16) +
          '. Got 0x' + manufacturerId.toString(16) +
          '. MCP9808 sensor not found.'
        ));
      }
      return i2cMcp9808.deviceId();
    }).then(deviceId => {
      if (deviceId !== DEVICE_ID) {
        return Promise.reject(new Error(
          'Expected device ID to be 0x' + DEVICE_ID.toString(16) +
          '. Got 0x' + deviceId.toString(16) +
          '. MCP9808 sensor not found.'
        ));
      }
      return i2cMcp9808.softReset();
    }).then(_ => {
      if (options.lowerAlertTemperature !== undefined) {
        return Promise.all([
          i2cMcp9808.lowerAlertTemperature(options.lowerAlertTemperature),
          i2cMcp9808.upperAlertTemperature(options.upperAlertTemperature),
          i2cMcp9808.criticalTemperature(options.criticalTemperature)
        ]);
      }
    }).then(_ => {
      if (options.hysteresis !== undefined) {
        return i2cMcp9808.hysteresis(options.hysteresis);
      }
    }).then(_ => {
      if (options.resolution !== undefined) {
        return i2cMcp9808.resolution(options.resolution);
      }
    }).then(_ => tempSensor);
  }

  close() {
    return Promise.resolve().then(_ => {
      if (this._alertGpio !== null) {
        this._alertGpio.unexport();
      }
      return this._i2cMcp9808.close();
    });
  }

  enableAlerts() {
    return this._i2cMcp9808.enableAlerts();
  }

  temperature() {
    return this._i2cMcp9808.temperature();
  }
}

module.exports = Mcp9808;

